extensions/funqy/funqy-knative-events/deployment/src/test/java/io/quarkus/funqy/test/WithAttributeFilter.java

```diff
@@ -0,0 +1,31 @@
+package io.quarkus.funqy.test;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import io.quarkus.funqy.Funq;
+import io.quarkus.funqy.knative.events.CloudEventMapping;
+import io.quarkus.funqy.knative.events.EventAttribute;
+import io.smallrye.mutiny.Uni;
+
+public class WithAttributeFilter {
```

Comment:
@mswiderski as I understand it, it is possible for filter on multiple attributes. Test cases should test that too (e.g. filter on source and subject).

Is dynamic dispatch supported? What I mean by dynamic dispatch: if there are two methods with different filters the runtime can pick automatically matching function (same way it works for trigger(type) now).

Reply:
@matejvasek test case for filtering by multiple attributes is here https://github.com/quarkusio/quarkus/pull/17655/files#diff-02c86bd7c06f51dbb63cbeb424137c6685d25ab51722469c3c175219bd9a6b44R66 it tests one standard CE attribute (`source`) and one extension attribute (`custom`)

yes, dynamic dispatch is there as function invokers are grouped by type and then filtered out based on matching attributes if any are defined.

Reply:
>test case for filtering by multiple attributes is here

sorry I missed that `@EventAttribute(name = "custom", value = "hello") })` in diff :man_facepalming: 

> yes, dynamic dispatch is there as function invokers are grouped by type and then filtered out based on matching attributes if any are defined.

Is this tested? I see two functions with different type (so dispatch is carried over it). Maybe there could be another function `toSemicolonSeparated()` with the same trigger as `toCommaSeparated()` but with different `source`.

Also is it possible that some filters are subset of another meaning that multiple functions match? If so which one is picked? The first declared?

Note I didn't review actual code. I am now just looking at test I trying to understand what it's supposed to do not how.

Reply:
> Is this tested? I see two functions with different type (so dispatch is carried over it). Maybe there could be another function toSemicolonSeparated() with the same trigger as toCommaSeparated() but with different source.

you're right there is no test covering that and I agree that it makes sense to have it in

> Also is it possible that some filters are subset of another meaning that multiple functions match? If so which one is picked? The first declared?

yes, there can be multiple functions matching and at the moment they are all invoked. This might be to optimistic as it only makes sense to invoke all if they are without output.
So I think we can do following

- invoke all if they don't have the output
- throw an error (409 - conflict) in case there are multiple functions matching and have outputs

what do you think?

Reply:
I don't like having different behavior for functions with and without output.

Maybe we could pick function with the most specific filter.
Example: if there is a filter on `source` and a filter on `source & subject`, the latter would be picked.
This assumes that functions with identical filter doesn't exist (it could be compilation error maybe?)
Also this approach might not work with the `CESQL` if we decided to implement it in future,
because detection whether some expression is a specialization of some other expression might be difficult/impossible.

So IMHO most simple solution would be outright forbid multiple matches (preferably in compile time).

@patriot1burke wdyt?

Reply:
@matejvasek I have updated tests to cover what we discussed and now there is also build check if there are no duplicates for the same type and filters.

Reply:
I am still not sure what should we do with non-equivalent but conflicting. Shall we also forbid that too? wdyt @patriot1burke

Reply:
@matejvasek could you please give an example of "non-equivalent but conflicting"?

Reply:
> if there is a filter on `source` and a filter on `source & subject`

I mean this

Reply:
thanks @matejvasek I will try to cover this case as well